src/components/SearchBar.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+
+export interface SearchBarProps {
+    onTextChange: Function;
+    onStructuredModeChanged: Function;
+    onDetailedModeChanged: Function;
+    onCategoriesChange: Function;
+    categories: string[];
+}
+
+export interface SearchBarState {
+    expanded: boolean;
+    selectedCategories: string[];
+    structured: boolean;
+    detailed: boolean;
+}
+
+export class SearchBar extends React.Component<SearchBarProps, SearchBarState> {
+
+    constructor(props: SearchBarProps) {
+        super(props);
+        this.state = { expanded: false, selectedCategories: this.props.categories, structured: false, detailed: false };
+    }
+
+    onTextChange(event: any) {
+        this.props.onTextChange(event);
+    }
+
+    onOptionsButtonClick() {
+        this.setState({ expanded: !this.state.expanded });
+    }
+
+    onStructuredModeChange(event: any) {
+        this.props.onStructuredModeChanged();
+        this.setState({ structured: !this.state.structured });
+    }
+
+    onDetailedModeChange(event: any) {
+        this.props.onDetailedModeChanged();
+        this.setState({ detailed: !this.state.detailed });
+    }
```

Comment:
Do we not need a parameter for `onStructuredModeChanged` and `onDetailedModeChanged` that carries the current value of `structured` and `detailed` values, respectively?